### ISSUE TITLE
`Var CodeLens` not always shown

### ISSUE THREAD

**Issue by BachiMjavanadze** [AdmiralSnyder/VarCodeLensExtension#3]
Sometimes `Var CodeLens` are not shown. Restarting the `Visual Studio IDE` partially solves the problem but not completely: some scripts show `Var CodeLens` and some don't.

![image](https://github.com/AdmiralSnyder/Support-VarCodeLens/assets/38082501/9fc85d6c-4f69-45b0-b7f1-521f41545e4a)

maybe the error lies in the following thing: if the scripts are open, then `Var CodeLens` does not always work. But if at the start of the `Visual Studio IDE` all the scripts are closed and we then open these scripts, then `Var CodeLens` , in my opinion, works... Honestly I can't catch the exact logic behind this error behavior

**Comment by BachiMjavanadze** [comment]
maybe the error lies in the following thing: if the scripts are open, then `Var CodeLens` does not always work. But if at the start of the `Visual Studio IDE` all the scripts are closed and we then run these scripts, then `Var CodeLens` , in my opinion, works... Honestly I can't catch the exact logic behind this error behavior

**Comment by AdmiralSnyder** [comment]
ah, i see. the problem most probably lies in me not getting a redraw event from VS - the extension works in two separate steps: First i create the vertical space to put the textboxes into (above the var keywords).
Then each occurrence of var is fetched and roslyn is asked for the type and i draw a textbox at the proper places.
but the latter operation is costly, so i do that only if VS tells me that the area has to be redrawn.
sadly, there's no "good" approach for getting notifications for that (or rather, there's lots of bad ones that are quite a lot of work to implement. so i'm listening to the "Layout Changed" event - which is invoked often _enough_ - but apparently not often enough for you. i'd recommend zooming in and out (using ctrl+mousewheel) if you find a situation where the type isn't rendered - that usually helps.

at some point i will work on this issue, in the future... but since VS has incorporated a feature to show the type of var (not above, though) - it doesn't have too hight a priority, right now, sorry.

**Comment by BachiMjavanadze** [comment]
> i'd recommend zooming in and out (using ctrl+mousewheel) if you find a situation where the type isn't rendered - that usually helps.

it helps partially; after zooming in and out (using ctrl+mouse wheel) I need to close script and reopen it again, otherwise `Var CodeLens` is not shown. it makes extension very hard to use

**Comment by Guiorgy** [comment]
Zooming in and out didn't help me. What did was scrolling away so the line with the var dissapears and then back in:

Before:
![image](https://github.com/AdmiralSnyder/Support-VarCodeLens/assets/27736965/877c5f62-4d2e-46d0-b5aa-ae1aacf7bbdb)
After:
![image](https://github.com/AdmiralSnyder/Support-VarCodeLens/assets/27736965/586b8316-5f4a-40c8-9c0c-b0f0695b9195)

But as you can see, although it gets redrawn, so do the others, and every time a redraw is supposed to happen, it may be missed.

**Comment by YourYoungerBrothersPug** [comment]
+1

**Comment by AdmiralSnyder** [comment]
yeah, i know, this is a bad situation - there's no api to be informed when e.g. github copilot outputs their stuff. i might be able to look into that in a couple weeks, busy with other stuff right now.
